src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,510 @@
+            : base(CreateMechanism(credential, properties, randomByteGenerator, dateTimeProvider))
+        {
+        }
+
+        internal MongoAWSAuthenticator(
+            string username,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock dateTimeProvider)
+            : base(CreateMechanism(username, null, properties, randomByteGenerator, dateTimeProvider))
+        {
+        }
+
+        /// <inheritdoc/>
+        public override string DatabaseName
+        {
+            get { return "$external"; }
+        }
+
+        // private static methods
+        private static MongoAWSMechanism CreateMechanism(
+            UsernamePasswordCredential credential,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock clock)
+        {
+            if (credential.Source != "$external")
+            {
+                throw new ArgumentException("MONGO AWS authentication may only use the $external source.", nameof(credential));
+            }
+
+            return CreateMechanism(credential.Username, credential.Password, properties, randomByteGenerator, clock);
+        }
+
+        private static MongoAWSMechanism CreateMechanism(
+            string username,
+            SecureString securePassword,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock clock)
+        {
+            var awsCredentialsCreators = new Func<AwsCredentials>[]
+            {
+                () => CreateAwsCredentialsFromMongoCredentials(username, securePassword, properties),
+                () => CreateAwsCredentialsFromEnvironmentVariables(),
+                () => CreateAwsCredentialsFromEcsResponse(),
+                () => CreateAwsCredentialsFromEc2Response()
+            };
+
+            foreach (var awsCredentialsCreator in awsCredentialsCreators)
+            {
+                var awsCredentials = awsCredentialsCreator();
+
+                if (awsCredentials == null)
+                {
+                    continue;
+                }
+
+                ValidateAwsCredentials(awsCredentials);
+
+                if (awsCredentials.Username == null)
+                {
+                    continue;
+                }
+
+                var credentials = new UsernamePasswordCredential("$external", awsCredentials.Username, awsCredentials.Password);
+
+                return new MongoAWSMechanism(credentials, awsCredentials.SessionToken, randomByteGenerator, clock);
+            }
+
+            throw new InvalidOperationException(AccessKeyIdExceptionMessage);
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromMongoCredentials(string username, SecureString securePassword, IEnumerable<KeyValuePair<string, string>> properties)
+        {
```

Comment:
by @rstam 
Should probably add:

    if (username == null)
    {
        return null;
    }

Either return a fully usable credential or return null.

Reply:
That's how it was previously implemented. I'm ok to change it back with @DmitryLukyanov confirmation.

Reply:
I like the currently implemented code since it allows calling ValidateAwsCredentials only once. So, the algorithm can be described as:
1. Create whatever credentials you want.
2. Then, validate the received credentials once.

But taking the retro discussion into account, I will rely on Mikalai to choose the right way to do it. 
After offline discussion, it sounds like the both suggestions can be mixed. Which is good :)